Sources/Algorithms/LazySplitSequence.swift

```diff
@@ -0,0 +1,321 @@
+//===----------------------------------------------------------------------===//
+//
+// This source file is part of the Swift Algorithms open source project
+//
+// Copyright (c) 2021 Apple Inc. and the Swift project authors
+// Licensed under Apache License v2.0 with Runtime Library Exception
+//
+// See https://swift.org/LICENSE.txt for license information
+//
+//===----------------------------------------------------------------------===//
+
+/// A sequence that lazily splits a base sequence into subsequences separated by
+/// elements that satisfy the given `whereSeparator` predicate.
+///
+/// - Note: This type is the result of
+///
+///     x.split(maxSplits:omittingEmptySubsequences:whereSeparator)
+///     x.split(separator:maxSplits:omittingEmptySubsequences)
+///
+///   where `x` conforms to `LazySequenceProtocol`.
+public struct LazySplitSequence<Base: Sequence> {
+  internal let base: Base
+  internal let isSeparator: (Base.Element) -> Bool
+  internal let maxSplits: Int
+  internal let omittingEmptySubsequences: Bool
+}
+
+extension LazySplitSequence {
+  public struct Iterator {
+    public typealias Element = [Base.Element]
+
+    internal var base: Base.Iterator
+    internal let isSeparator: (Base.Element) -> Bool
+    internal let maxSplits: Int
+    internal let omittingEmptySubsequences: Bool
+    internal var separatorCount = 0
+    internal var sequenceLength = 0
+
+    internal init(
+      base: Base.Iterator,
+      whereSeparator: @escaping (Base.Element) -> Bool,
+      maxSplits: Int,
+      omittingEmptySubsequences: Bool
+    ) {
+      self.base = base
+      self.isSeparator = whereSeparator
+      self.maxSplits = maxSplits
+      self.omittingEmptySubsequences = omittingEmptySubsequences
+    }
+  }
+}
+
+extension LazySplitSequence.Iterator: IteratorProtocol {
+  public mutating func next() -> Element? {
+    /// Separators mark the points where we want to split (cut in two) the base
```

Comment:
This seams like a doc comment, should it be above the function instead of the body?

Reply:
This is not a doc comment, but rather a high-level discussion (more general than any particular implementation) of how the split operation works. I found writing this to be very helpful to my understanding of how any split implementation should work. In retrospect, it might serve better as the introduction to the /Guides/Split.md I need to write. I can see how in its current location its intention is unclear.

Reply:
Ah I see, thanks! I mention doc comments because comments starting with `///` are meant for documentation :)
But I agree such detailed info about algorithm itself could be on the guides. WDYT @natecook1000 @timvermeulen?

Reply:
I used `///` because when I'm mixing prose and code, I like seeing each rendered differently as Xcode does with `///`. Should that style only be used for comments intended for the generated documentation?

Reply:
Yes, `///` is for documentation comments.